ENH: extend factorial{,2,k} to allow complex inputs

h-vetinari opened the pull request:
Main implementation for #18409; builds on #19989 and #21702

Closes #18409

Note that there's a lot of simplification possible once the implementation lands, see #19988

h-vetinari committed:
MAINT: prepare tests for `extend=` kwarg of factorial{,2,k}

h-vetinari committed:
MAINT: remove helpmsg from factorialk errors; will reach parity with factorial{,2}

h-vetinari committed:
ENH: extend factorial{,2,k} to allow complex inputs

h-vetinari committed:
MAINT: also test infinities for factorial{,2,k}

h-vetinari committed:
TST: ignore test_warning_calls_filters for special/_basic.py

we legitimately want to ignore some warnings from NumPy that are unhelpful here

steppi commented on scipy/special/_basic.py:
```diff
@@ -2911,34 +2911,66 @@ def _factorialx_array_exact(n, k=1):
     """
     Core approximation to multifactorial for array n and integer k.
     """
+    def _gamma_no_inf(vals):
+        res = gamma(vals)
+        # replace infinities coming from gamma function with nan, see #19071;
+        # gamma only returns inf for real inputs, so no fix needed for complex inputs
+        if isinstance(res, np.ndarray):
+            if not _is_subdtype(vals.dtype, "c"):
+                # we _do_ keep infinity where input itself was +∞
+                keep_as_inf = np.isinf(vals) & (vals > 0)
+                res[np.isinf(res) & ~keep_as_inf] = np.nan
+        elif np.isinf(res) and not (np.isinf(vals) and (vals > 0)):
+            res = np.float64("nan")
+        return res
```
@h-vetinari, at this point, I think we should just fix `gamma` and related functions. Hopefully there aren't actually too many users relying on this bug in the behavior at the poles, but if so, it should be fine to just add a note in the docstring explaining the old and new behavior, and to use `rgamma` instead of dividing by the `gamma` function.

I'll make an issue and discourse post and see if we can get the fix through this week.

Sorry, I was the one dragging my feet on fixing this for fear of breaking someones workflow, but seeing the workarounds you had to do has convinced me.

I'll make an issue about it and see if we can get this through

h-vetinari replied:
I had completely forgotten about the discussion in #19071 already - that would probably be a good candidate for the issue you're talking about? Happy to switch to `rgamma` of course.

h-vetinari replied:
Fixed as discussed in #21827

h-vetinari committed:
MAINT: simplify work-around for gamma infinities

following https://github.com/scipy/scipy/commit/e5b0af896ef19285677e891f23d338d0b3ba16ed

steppi reviewed (commented):
Thanks @h-vetinari. This looks really good. No changes requested implementation wise. A couple comments and questions regarding docstrings and warning messages.

h-vetinari commented on scipy/special/_basic.py:
```diff
@@ -3170,6 +3265,17 @@ def factorialk(n, k, exact=None):
+
+        .. warning::
+
+           Using the ``'complex'`` extension also changes the values of the
+           multifactorial at integers ``n != 1 (mod k)`` by a factor depending
+           on both ``k`` and ``n % k``, see above or [1].
```
This is a leftover from reshuffling (I've already fixed it locally but didn't want to repush again for a single word change), it should say "below". The factor in question can be seen towards the bottom of the docstring

steppi replied:
I see. That makes sense.  Just do a `[docs only]` commit with any docstring changes needed, and then I think this is ready to merge.

h-vetinari replied:
> Just do a `[docs only]` commit with any docstring changes needed, and then I think this is ready to merge.

I'd actually prefer to keep the commit history here. I'll push a rebase with the typos you found - thank you for the review!

steppi replied:
Sure, no problem. Commit history looks clean.

h-vetinari commented:
I rebased on main, the failures must be coming from there... I don't think that needs to hold up this PR? If you agree, I'll let you approve and merge @steppi 🙏

steppi commented:
> I rebased on main, the failures must be coming from there... I don't think that needs to hold up this PR? If you agree, I'll let you approve and merge @steppi 🙏

Yes, failures look unrelated. I'll merge if CI settles with no related failures. Thanks

steppi reviewed (approved):
lgtm
